pkg/services/accesscontrol/accesscontrol.go

```diff
@@ -15,6 +15,14 @@ type AccessControl interface {
 
 	// Middleware checks if service disabled or not to switch to fallback authorization.
 	IsDisabled() bool
+
+	// AddFixedRoleRegistrations allow the caller to declare, to the
+	// service, fixed roles and their assignments
+	AddFixedRoleRegistrations(...RoleRegistration)
```

Comment:
These interfaces are a bit confusing to be honest. `add` and `register` have very similar semantics, what could we name those to reflect better the meaning? It is also hard to understand what "launch all registrations" mean? 

What I understand is that `AddFixedRoleRegistrations` simply appends to the fixed role list and `RegisterFixedRoles` validates and persist whatever is in memory. 
- I think validation of the fixed role should happen at the registration time, not when we persist. So basically when calling `AddFixedRoleRegistrations`
- `RegisterFixedRoles` can only persist whatever configuration we have. That said, currently a client needs to remember to persist registered roles, which is leaves gaps and I'd say that persisting roles should not be a clients responsibility, but rather be isolated within access control package. So the question is, if we can remove `RegisterFixedRoles` from this interface at all and make sure that `accesscontrol` persists all roles registered through `AddFixedRoleRegistrations`?

Reply:
The idea was to follow the SQL AddMigration pattern:
Packages call AddMigration, then Migrate - the function that actually performs the migration - happens only once server wise. So to keep the same semantics and behavior, I used: ~AddRegistrations (called by all packages) and ~Register (called once server wise).
Would you find it easier to understand with ~AppendRegistration and ~Register?
To give a little bit of history on the matter, at the beginning we only had one single method that would perform the registration, but since the AccessControl service wasn't the first service initialized, services with higher priorities registering roles would panic the server.

Reply:
Thanks for the context, it helps. In the AddMigration patter though, the actual migrations are executed from one single place or (in `sqlstore.go` from what I see)? So do you mean that noone is expected to call `RegisterFixedRoles` , beside having it in the server init, like you have now (as what is confusing to me is that the interface exposes a method which in theory makes it free for anyone to call from anywhere)?

Regarding the naming, from my point of view, if we can name methods simply based on what they do - e.g. `AddFixedRole` and `PersistFixedRoles` , that would be easiest to understand. Also I still think that it would be good to do validations when adding a fixed role, not when persisting.

Reply:
Alright, this makes sense.
What do you think about this:
```go
type AccessControl interface {
	// Evaluate evaluates access to the given resource.
	Evaluate(ctx context.Context, user *models.SignedInUser, permission string, scope ...string) (bool, error)

	// GetUserPermissions returns user permissions.
	GetUserPermissions(ctx context.Context, user *models.SignedInUser) ([]*Permission, error)

	// Middleware checks if service disabled or not to switch to fallback authorization.
	IsDisabled() bool

	// RequestRoleRegistrations allow the caller to declare, to the
	// service, fixed roles and their assignments
	RequestRoleRegistrations(...RoleRegistration) error
}
```
Maybe `RequestRoleRegistrations` is a bit better than persist, because we don't actually persist anything in OSS. I also hope, it is clear enough to show that this is an async process? Note that I also added an error for validation of the name and assignments.

In server.go we can define another interface that the AccessControl services would implicitly fullfill:
```go
type RoleRegistry interface {
	// RegisterFixedRoles process all registration requests
	RegisterFixedRoles() error
}
```

Reply:
👍  for the interface separation. About the naming, `AddFixedRole` is the most intuitive one to be honest, especially if we separate into two different interfaces, the meaning will be also carried over from the interface name, so we can easily have `add` for adding to the stack and `register` for actually executing. but I don't want to stuck forever on naming discussion, so please feel free to go with whatever you find best 👍